### ISSUE TITLE
Order of scalar variables with respect to structured naming

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/fmi-standard#413]
**Reported by henning&#46;larsson on 9 Mar 2017 11:11 UTC**
The FMI 2.0 standard specifies the ordering of structured variable names in modelDescription.xml (p. 65):

 "The tree of names is mapped to an ordered list of ScalarVariable.name’s  in depth-first order. Example:
```
 vehicle
   transmission
     ratio
     outputSpeed
   engine
     inputSpeed
     temperature
```
 is mapped to the following list of ScalarVariable.name’s:
```
 vehicle.transmission.ratio
 vehicle.transmission.outputSpeed
 vehicle.engine.inputSpeed
 vehicle.engine.temperature
```
 All array elements are given in a consecutive sequence of  ScalarVariables. Elements of multidimensional arrays are ordered according  to “row major” order that is elements of the last index are given in sequence."

FMI 1.0 has a similar definition.

Currently, the compliance checker verifies the content of structured names, but not the ordering. If a tool generates a modelDescription from the example above with ordering
```
 vehicle.transmission.ratio
 vehicle.engine.inputSpeed
 vehicle.transmission.outputSpeed
 vehicle.engine.temperature
```
it would be incorrect, but pass the compliance checker (version 2.0.3) without warnings or errors.

----

Migrated-From: https://trac.fmi-standard.org/ticket/413

**Comment by modelica-trac-importer** [comment]
**Comment by maplesoft on 26 Feb 2018 20:09 UTC**
The specification document states (on page 65):

"The tree of names is mapped to an ordered list of ScalarVariable.name’s in depth-first order."

But this is the ‘tree’ as implemented by the importing tool.  It is our interpretation that this
only dictates how to construct the names from the tree structure.  It does not specify
the order in which the ScalarVariables must appear in the modelDescription.xml file.

It would be unnecessarily restrictive to force the xml to list all the ScalarVariables in such order.
For example, we export FMUs for tools that require the variables to be ordered by causality: 
independent, input, output, parameter, etc. The proposed limitation would prevent them from 
using structured names. 

The specification does state that the array elements must be given in a consecutive sequence
but not all the other names.  We’re not sure what the benefit of the ordering restriction would
be other than "ease of implementation" for the importing tool.

**Comment by modelica-trac-importer** [comment]
**Comment by maplesoft on 26 Feb 2018 20:31 UTC**
Also, the example of a modelDescrption.xml on pg. 91-92 of the spec lists:
variableNamingConvention="structured"
but the names appearing in the ModelVariables section are not ordered, and there's no need for them to be.

**Comment by modelica-trac-importer** [comment]
**Comment by karl&#46;wernersson on 27 Feb 2018 10:40 UTC**
I agree with Henning that this is a bug, that does need to be fixed.
I disagree with maplesoft and thought that the specification is sufficient clear, however since we have some disagreement a clarification might be good for the FMI 2.0.1 release.
Also as maplesoft pointed out the example on pg.91-92 is wrong in this regard and needs to be updated.
If you are not able to generate an xml accordingly, the flat option exists.
In the end, we want nice xml files and it would be nice if the importer don't have to rearrange and sort all scalars variables but rather use the declared order in the xml

**Comment by modelica-trac-importer** [comment]
**Comment by cbertsch on 30 May 2018 14:42 UTC**
Moved to https://github.com/modelica-tools/FMUComplianceChecker/issues/28

**Comment by modelica-trac-importer** [comment]
**Comment by cbertsch on 1 Jun 2018 09:13 UTC**
Reopened, as a clarification in the standard seem necessary and this is not only a compliance checker topic. See the discussion in https://github.com/modelica-tools/FMUComplianceChecker/issues/28

**Comment by KarlWernersson** [comment]
In my view this is clear and don't need an update.
Except for the example on page 91-92
in addition the dcp standard
https://github.com/modelica/dcp-design
has also choose to go this way.
We still have the flat option for any one unable to follow it,

**Comment by KarlWernersson** [comment]
Poll:
Relax definition not to take care about ordering:
Antoine, Christian Winther, Markus, Massoud, Otto, Thomas, Torsten
Interpret have a tought definition where the ordering is fixed:
Karl.
Abstain: Adreas, Irina, Christian Bertsch, kalus, Timo.

**Comment by chrbertsch** [comment]
Included in the merged PR #579
